compiler-core/src/reference.rs

```diff
@@ -44,51 +80,243 @@ impl ReferenceTracker {
+        self.current_function = self.get_or_create_node(name, EntityLayer::Value);
+    }
+
+    pub fn register_type(
+        &mut self,
+        name: EcoString,
+        kind: EntityKind,
+        location: SrcSpan,
+        publicity: Publicity,
+    ) {
+        self.current_function = self.create_node(name.clone(), EntityLayer::Type);
+        let entity = Entity {
+            name,
+            layer: EntityLayer::Type,
+        };
+        match publicity {
+            Publicity::Public | Publicity::Internal { .. } => {
+                self.public_entities.push(entity.clone());
+            }
+            Publicity::Private => {}
+        }
+
+        _ = self.entity_information.insert(
+            entity,
+            EntityInformation {
+                kind,
+                origin: location,
+            },
+        );
+    }
+
+    pub fn register_module(&mut self, name: EcoString, kind: EntityKind, location: SrcSpan) {
+        self.current_function = self.create_node(name.clone(), EntityLayer::Module);
+        let entity = Entity {
+            name,
+            layer: EntityLayer::Module,
+        };
+
+        _ = self.entity_information.insert(
+            entity,
+            EntityInformation {
+                kind,
+                origin: location,
+            },
+        );
     }
 
     pub fn register_value_reference(
         &mut self,
         module: EcoString,
         name: EcoString,
+        referenced_name: &EcoString,
         location: SrcSpan,
         kind: ReferenceKind,
     ) {
-        let target = self.get_or_create_node(module.clone(), name.clone());
+        match kind {
+            ReferenceKind::Qualified | ReferenceKind::Import | ReferenceKind::Definition => {}
+            ReferenceKind::Alias | ReferenceKind::Unqualified => {
+                let target = self.get_or_create_node(referenced_name.clone(), EntityLayer::Value);
+                _ = self.graph.add_edge(self.current_function, target, ());
+            }
+        }
+
         self.value_references
             .entry((module, name))
             .or_default()
             .push(Reference { location, kind });
-
-        _ = self.graph.add_edge(self.current_function, target, ());
     }
 
     pub fn register_type_reference(
         &mut self,
         module: EcoString,
         name: EcoString,
+        referenced_name: &EcoString,
         location: SrcSpan,
         kind: ReferenceKind,
     ) {
+        match kind {
+            ReferenceKind::Qualified | ReferenceKind::Import | ReferenceKind::Definition => {}
+            ReferenceKind::Alias | ReferenceKind::Unqualified => {
+                self.register_type_reference_in_call_graph(referenced_name.clone())
+            }
+        }
+
         self.type_references
             .entry((module, name))
             .or_default()
             .push(Reference { location, kind });
     }
+
+    pub fn register_type_reference_in_call_graph(&mut self, name: EcoString) {
```

Comment:
What's the difference between this and register_type_reference? It's not clear to me why one is used some times and the other other times.